api/src/main/java/org/apache/iceberg/FileScanTask.java

```diff
@@ -47,6 +47,13 @@
    */
   PartitionSpec spec();
 
+  /**
+   * The partition data for the file of this task.
```

Comment:
What does this return if there is no partition?

I think that we should consider adding a struct type that describes this tuple. That way, we can use an empty struct for unpartitioned and a non-empty struct for tasks that are combined by partition. We could also support more combinations, like combining across day partitions by not across bucket partitions or similar.

Reply:
We probably also want to expose the struct type at the scan level so that we have a consistent type across all tasks, instead of adding it to each task.

Reply:
> What does this return if there is no partition?

It will return `BaseFile.EMPTY_PARTITION_DATA`, see changes to `StaticDataTask.java` for example.

> I think that we should consider adding a struct type that describes this tuple. That way, we can use an empty struct for unpartitioned and a non-empty struct for tasks that are combined by partition. We could also support more combinations, like combining across day partitions by not across bucket partitions or similar.

Hmm isn't `spec()` above designed for this purpose?

Reply:
@rdblue, are you suggesting we should build a unified partition type in case if multiple specs are present?

Reply:
I am not sure I fully understood.

Reply:
What I'm thinking is that you might want to only restrict combining by one or two partition fields. For example, say the query's join condition is `a.id = b.id` and table `a` is partitioned by `day(a.ts), bucket(128, a.id)`. Then it's fine to combine splits in different date partitions, just not in different buckets. In that case, the splits that we produce should have a bucket partition field, but not a date partition field.

It's a little confusing that I made this comment on `FileScanTask`, which is always in a single partition. But the `Task` API still has the same concern: we may want to avoid combining in a subset of partitions and those are the values that we return through a `partition` method.

Reply:
This makes sense. It requires Spark to pushdown the info to data sources like Iceberg though, and I'm not sure how complex that would be. My current approach is to do this in Spark, by pushing down the join/aggregate keys to scan node where it combines the input partitions accordingly.

Reply:
Yeah, for now I think it's a good idea to not combine across any partition boundary. Later, we may want to, though. So as long as we're looking at the API I think this is something we should consider.